Tutorial - Ruby on rails ab tests (#7790)
PostHog/posthog.com: contents/tutorials/ruby-ab-tests.md

Author: I'll rename to `ruby-on-rails-ab-tests`, since I also want to target ruby ab testing (for folks looking for backend)

---
title: How to set up A/B tests in Ruby on Rails
date: 2024-02-13
author: ["lior-neu-ner"]
tags: ['experimentation']
---

import { ProductScreenshot } from 'components/ProductScreenshot'
import EventsInPostHogLight from '../images/tutorials/ruby-ab-tests/events-light.png'
import EventsInPostHogDark from '../images/tutorials/ruby-ab-tests/events-dark.png'
import TestSetupLight from '../images/tutorials/ruby-ab-tests/experiment-setup-light.png'
import TestSetupDark from '../images/tutorials/ruby-ab-tests/experiment-setup-dark.png'

A/B tests help you improve your Ruby on Rails app by enabling you to compare the impact of changes on key metrics. 

To show you how to set one up on both the client and server side, we create a basic Rails app, add PostHog, create an A/B test, and implement the code for it.

## 1. Create a Rails app

> First, ensure [Ruby](https://www.ruby-lang.org/en/documentation/installation/) and [Rails](https://guides.rubyonrails.org/v5.0/getting_started.html) are both installed. Then, create a new Rails app:

Reviewer: This process can be a bit frustrating so might be helpful to help a bit more.
```suggestion
First, ensure [Ruby](https://www.ruby-lang.org/en/documentation/installation/) and [Rails](https://guides.rubyonrails.org/v5.0/getting_started.html) are both installed ([rbenv](https://github.com/rbenv/rbenv) is a useful tool for using a compatible Ruby version). Then, create a new Rails app:
```

Author: I dont think we need to do this. People who are reading this post are likely well-versed in Ruby and already set up,

```bash
rails new ruby-ab-tests
cd ruby-ab-tests
```

Next, we set up a home page with a simple heading and button. To do this, generate a `Pages` controller with a `home` action:

```bash
rails generate controller Pages home
```

Then replace the code in `app/views/pages/home.html.erb` with the following:

```html file=app/views/pages/home.html.erb
<main>
  <h1>Rails A/B Tests</h1>
  <button id="main-cta">Click me</button>
</main>
```

Lastly, modify `config/routes.rb` to make our new home page the root path of our app:

```ruby file=config/routes.rb
Rails.application.routes.draw do
  root 'pages#home'
end
```

Run `rails server` and navigate to `http://localhost:3000` to see our app in action.

![Basic Ruby app](../images/tutorials/ruby-ab-tests/basic-app.png)

## 2. Add PostHog on the client side

With our app set up, it’s time to install and set up PostHog. If you don't have a PostHog instance, you can [sign up for free](https://us.posthog.com/signup).

> To start, we install the [PostHog Web SDK](/docs/libraries/js) on the client side using the `Web snippet`. Copy the snippet from [your project settings](https://us.posthog.com/settings/project#snippet) and paste it in the `<head>` tag of `views/layouts/application.html.erb`:

Reviewer: ```suggestion
To start, we install the [PostHog Web SDK](/docs/libraries/js) on the client side using the `Web snippet`. Copy the snippet from [your project settings](https://us.posthog.com/settings/project#snippet) and paste it in the `<head>` tag of `app/views/layouts/application.html.erb`:
```

> ```html views/layouts/application.html.erb

Reviewer: ```suggestion
```html app/views/layouts/application.html.erb
```

<!DOCTYPE html>
<html>
  <head>

>     <title>RubyAbTests2</title>

Reviewer: ```suggestion
    <title>RubyAbTests</title>
```

    <%= javascript_importmap_tags %>
    
    <!-- PostHog Snippet --> 
    <script>
      !function(t,e){var o,n,p,r;e.__SV||(window.posthog=e,e._i=[],e.init=function(i,s,a){function g(t,e){var o=e.split(".");2==o.length&&(t=t[o[0]],e=o[1]),t[e]=function(){t.push([e].concat(Array.prototype.slice.call(arguments,0)))}}(p=t.createElement("script")).type="text/javascript",p.async=!0,p.src=s.api_host+"/static/array.js",(r=t.getElementsByTagName("script")[0]).parentNode.insertBefore(p,r);var u=e;for(void 0!==a?u=e[a]=[]:a="posthog",u.people=u.people||[],u.toString=function(t){var e="posthog";return"posthog"!==a&&(e+="."+a),t||(e+=" (stub)"),e},u.people.toString=function(){return u.toString(1)+".people (stub)"},o="capture identify alias people.set people.set_once set_config register register_once unregister opt_out_capturing has_opted_out_capturing opt_in_capturing reset isFeatureEnabled onFeatureFlags getFeatureFlag getFeatureFlagPayload reloadFeatureFlags group updateEarlyAccessFeatureEnrollment getEarlyAccessFeatures getActiveMatchingSurveys getSurveys onSessionId".split(" "),n=0;n<o.length;n++)g(u,o[n]);e._i.push([i,s,a])},e.__SV=1)}(document,window.posthog||[]);

>       posthog.init('<ph_project_api_key>',{api_host:'ph_instance_address})

Reviewer: ```suggestion
      posthog.init('<ph_project_api_key>',{api_host:'<ph_instance_address>'})
```

Next, we create an initializer for our PostHog client. Create a new file in `config/initializers` called `posthog.rb`. Then, initialize PostHog in this file using your project API key and host:

```ruby file=config/initializers/posthog.rb
require 'posthog-ruby'

POSTHOG_CLIENT = PostHog::Client.new({
  api_key: '<ph_project_api_key>',
  host: '<ph_instance_address>'
})
```

Then, we fetch the feature flag in our controller `app/controllers/pages_controller.rb` and set the button text using the value from it:

```ruby file=app/controllers/pages_controller.rb
class PagesController < ApplicationController
  def home
    @button_text = 'No variant'
    begin
      distinct_id = 'placeholder-user-id'
      enabled_variant = POSTHOG_CLIENT.get_feature_flag('my-cool-experiment', distinct_id)
      if enabled_variant == 'control'
        @button_text = 'Control Variant'
      elsif enabled_variant == 'test'
        @button_text = 'Test Variant'
      end
    rescue => e
      @button_text = 'Error'
    end
  end
end
```

Lastly, update our view to use the `@button_text` variable from the controller:

```html file=app/views/pages/home.html.erb
<script>
  function handleClick() {

>     console.log(posthog)

Reviewer: ```suggestion
```

    posthog.capture('home_button_clicked');
  }
</script>

<main>
  <h1>Rails A/B Test</h1>
  <button id="main-cta" onclick="handleClick()"><%= @button_text %></button>
</main>
```

Now, when you refresh the page, the button text is already set when the page loads. 

> 

Reviewer: Since we use both the server and client side here, this A/B test isn't going to work properly because the backend ID is going to be different than the frontend ID. I realize we partially solve this in the next section, but maybe we can add a bit more code to pass the server side ID (as a variable) and call identify on the frontend? 

I don't want to end this section with a broken implementation.

Author: Added new section on this

#### Setting the correct `distinctId`

You may notice that we set `distinctId = 'placeholder-user-id'` in our flag call above. In production apps, to ensure you fetch the correct flag value for your user, `distinctId` should be set to their unique ID. 

For logged-in users, you typically use their email as their `distinctId`. However, for logged-out users, you can use the `distinct_id` property from their PostHog cookie:

> 

Reviewer: ```suggestion
```

```ruby file=app/controllers/pages_controller.rb
class PagesController < ApplicationController
  def home
    @button_text = 'No variant'
    
    project_api_key = '<ph_project_api_key>'
    cookie_key = "ph_#{project_api_key}_posthog"

    if cookies[cookie_key]
      begin
        distinct_id = JSON.parse(cookies[cookie_key])['distinct_id']
        enabled_variant = POSTHOG_CLIENT.get_feature_flag('my-cool-experiment', distinct_id)
        if enabled_variant == 'control'
          @button_text = 'Control variant'
        elsif enabled_variant == 'test'
          @button_text = 'Test variant'
        end
      rescue => e
        @button_text = 'Error'
      end
    end
  end
end
```

> In the scenario where your PostHog cookie is not available on the server side, you can fall back to the client side rendering approach of feature flags

Reviewer: Could we add the code that checks `@button_text` on the client side and checks the flag if it doesn't exist? 
```suggestion
In the scenario where your PostHog cookie is not available on the server side, you can fall back to the client side evaluation of the feature flag.
```
